staging/src/k8s.io/component-helpers/nodedeclaredfeatures/framework.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodedeclaredfeatures
+
+import (
+	"fmt"
+	"slices"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/version"
+)
+
+// Framework provides functions for discovering node features and inferring pod feature requirements.
+// It is stateful and holds the feature registry.
+type Framework struct {
+	registry []Feature
+}
+
+// FeatureSet is a set of node features.
+type FeatureSet struct {
+	sets.Set[string]
+}
+
+// NewFeatureSet creates a FeatureSet from a list of feature names.
+func NewFeatureSet(features []string) FeatureSet {
+	return FeatureSet{Set: sets.New(features...)}
+}
+
+// Equal returns true if both the sets have the same features.
+func (s *FeatureSet) Equal(other FeatureSet) bool {
+	return s.Set.Equal(other.Set)
+}
+
+// New creates a new instance of the Framework.
+func New(registry []Feature) (*Framework, error) {
+	if registry == nil {
+		return nil, fmt.Errorf("registry must not be nil")
+	}
+	return &Framework{
+		registry: registry,
+	}, nil
+}
+
+// DiscoverNodeFeatures determines which features from the registry are enabled
+// for a specific node configuration. It returns a sorted, unique list of feature names.
+func (f *Framework) DiscoverNodeFeatures(cfg *NodeConfiguration) []string {
```

Comment:
Should DiscoverNodeFeatures care about version? If so, I guess it would care about the node version, so maybe node version should be part of the NodeConfiguration struct?

Reply:
Makes sense. Added MaxVersion() check in DiscoverNodeFeatures().